packages/flutter/test/painting/matrix_utils_test.dart

```diff
@@ -121,4 +121,79 @@ void main() {
       forcedOffset,
     );
   });
+
+  test('transformRect with no perspective (w = 1)', () {
```

Comment:
Perhaps it would give us a nice sanity check to validate the results of this against the implementations from vector_math.  I trust that you did the right thing here, but it would make reviewing/changing this easier.

We might also want to consider upstreaming this into vector_math.

Reply:
_vectorWiseTransformRect used in the other (perspective) test uses vector_math. So, which of the following would make sense:

1 - MatrixUtils vs. _vectorWise/aka vector_math
2 - MatrixUtils vs. manually computed answers
3 - _vectorWise vs. manually computed answers

We currently do #2, are you suggesting replacing it with #1, or adding #1 (or #3)?

Reply:
Sorry, I missed where you were doing this.  Resolved.

Reply:
This comment is in a test file but you mention "upstreaming this into vector_math". Do you mean the test or the optimization (or both)?

The problem with moving the optimization upstream is that we use different classes to store our data than vector_math (except for Matrix4) so to avoid unpacking/repacking/moving the data to a different object, ...? This is why I inlined these calculations to avoid transfer of data for a common operation.

Note that vector_math has its own bounding objects, but they are much more abstract and complicated than our Rect.

Reply:
I don't see upstreaming as a blocker to this PR - more that this approach seems to be an improvement, and it might be something that would be valuable to have available via vector_math rather than burried in Flutter.

I think we can land this without doing that - if I'm not missing something here, it would be enough to just open a bug on vector_math to track making this available there.

Reply:
I did think of this as I was doing it and I did look through the code in vector_math, but in the end I realized that our specific needs aren't a great match for the tone of their API. In particular, the bounds objects they provide feel more oriented to the kind of math that a 3D CAD or game package would require and doesn't seem to speak the language of rectangles laid out on a screen.

Another thing to consider is how big vector_math is and how good our tree shaking is. Vector4, for example, has what looks like several hundred properties on it. Do we end up with getter code for all of those properties in a flutter app that is just trying to figure out which cat video someone poked with their finger?

This is a larger discussion than just this one fix. I'd almost lean towards creating our own matrix package that is more streamlined and tailored to our needs, but I believe we have Matrix4 baked into our APIs?

Reply:
From what I understand, vector_math _is_ the package we created to have matrix math available in Flutter, though it's certainly got other uses as well and has some other large contributions outside of what Flutter needs.

Reply:
Hmmm, well it doesn't provide any Rect-like object so this would require more than just adding a new method. And then we deal with data ferrying between our Rect and its Rect or we make a breaking change and move our Rect to the vector_math package?

Reply:
We can't move our Rect, that's defined in dart:ui which isn't allowed to have dependencies.

I'm not sure if there's a really good way to do this, but as I said It shouldn't block this PR from landing either way.

Reply:
I started at looking at how hard it would be to do just the Point version of this inside vector_math and came away not happy with the state of its APIs (consistency, capability, etc.).

Sad bear... :(